File: app/src/main/java/com/taskmanager/task/TaskManager.java
package com.taskmanager.task;

import android.util.SparseArray;

/**
 * Created by alexeyglushkov on 20.09.14.
 */

public interface TaskManager extends TaskPool {
    // Task Limits
    void setMaxLoadingTasks(int maxLoadingTasks);
    int getMaxLoadingTasks();

    void setLimit(int taskType, float availableQueuePart);
    SparseArray<Float> getLimits();
    SparseArray<Integer> getUsedSpace(); //type -> task count from loadingTasks



    // Task Running

    // Put a task, the same task can't be putted twice to the TaskManager
    // TODO: handle this situation well
    void startImmediately(Task task);
    void cancel(Task task, Object info);
    int getLoadingTaskCount();

    // Task Providers
    void addTaskProvider(TaskProvider provider);
    void removeTaskProvider(TaskProvider provider);
    SortedList<TaskProvider> getTaskProviders(); //always sorted by priority

    TaskProvider getTaskProvider(String id);
    void setTaskProviderPriority(TaskProvider provider, int priority);

    // Executor
    void setTaskExecutor(TaskExecutor executor);
    TaskExecutor getTaskExecutor();

    // Listeners
    void removeListener(TaskManagerListener listener);
    void addListener(TaskManagerListener listener);

    interface TaskManagerListener {
        void onLimitsChanged(TaskManager taskManager, int taskType, float availableQueuePart);

        // add additional flag showing that a task added to a loading or waiting queue
        void onTaskAdded(TaskPool pool, Task task, boolean isLoadingQueue);
        void onTaskRemoved(TaskPool pool, Task task, boolean isLoadingQueue);
    }
}

File: taskmanager/src/main/java/com/example/alexeyglushkov/taskmanager/task/TaskManager.kt
package com.example.alexeyglushkov.taskmanager.task

import android.util.SparseArray

import androidx.collection.SparseArrayCompat
import kotlinx.coroutines.CoroutineScope

import java.util.ArrayList

/**
 * Created by alexeyglushkov on 20.09.14.
 */

interface TaskManager : TaskPool, TaskPool.Listener {
    // TaskManagerCoordinator sets taskFilters to taskProviders
    var taskManagerCoordinator: TaskManagerCoordinator

    fun getLoadingTaskCount(): Int
    val taskProviders: List<TaskProvider> //always sorted by priority

    // Task Running
    // Put a task, the same task can't be putted twice to the TaskManager
    // TODO: handle this situation well
    fun startImmediately(task: Task)
    fun cancel(task: Task, info: Any?)

    // Task Providers
    fun addTaskProvider(provider: TaskProvider)
    fun removeTaskProvider(provider: TaskProvider)
    fun getTaskProvider(id: String): TaskProvider?
    fun setTaskProviderPriority(provider: TaskProvider, priority: Int)

    // Listeners
    fun removeListener(listener: Listener)
    fun addListener(listener: Listener)

    interface Listener {
        fun onLimitsChanged(taskManager: TaskManager, taskType: Int, availableQueuePart: Float)

        // add additional flag showing that a task added to a loading or waiting queue
        fun onTaskAdded(pool: TaskPool, task: Task, isLoadingQueue: Boolean)
        fun onTaskRemoved(pool: TaskPool, task: Task, isLoadingQueue: Boolean)
    }
}
